### PR TITLE
add zso support

### PR DESCRIPTION
**Pull request by orcutt989**
### Description of Changes

Adds ability to play ZSO files.

### Rationale behind Changes

This is a popular format that has been in development and many people are switching to this format over CSO and ISO.
https://github.com/PCSX2/pcsx2/issues/6585
https://github.com/PCSX2/pcsx2/issues/6055
https://github.com/PCSX2/pcsx2/issues/6050

### Suggested Testing Steps

Build and run
Or try this compiled build https://www.mediafire.com/file/vl0zc6g6886s63m/pcsx2-with-zso-support.7z/file

### Notes
Used @JoseAaronLopezGarcia's CPP file here https://github.com/PCSX2/pcsx2/issues/6050#issuecomment-1126741880 and made it work.  

Used https://wiki.pcsx2.net/PCSX2_Documentation/Compiling_on_Windows only to find out the NIS files had been removed and ended up using CMAKE.

### CODE REVIEW

**Comment by orcutt989** [comment]
@JoseAaronLopezGarcia I fixed up your CPP file and struggled with Visual Studio, and got a working build, but when I load a ZSO file it just does nothing.

Here is the build - https://file.io/wsPRz35mv4M3

File: pcsx2/CDVD/CsoFileReader.cpp [hunk 1]
```diff
@@ -160,14 +170,16 @@ bool CsoFileReader::InitializeBuffers()
...
 
-	m_z_stream = new z_stream;
-	m_z_stream->zalloc = Z_NULL;
-	m_z_stream->zfree = Z_NULL;
-	m_z_stream->opaque = Z_NULL;
-	if (inflateInit2(m_z_stream, -15) != Z_OK)
-	{
-		Console.Error("Unable to initialize zlib for CSO decompression.");
-		return false;
+	if (!m_uselz4){
```
Comment (lightningterror) [inline pcsx2/CDVD/CsoFileReader.cpp:173]: Formatting.

File: pcsx2/CDVD/CsoFileReader.cpp [hunk 2]
```diff
@@ -252,14 +264,21 @@ int CsoFileReader::ReadChunk(void *dst, s64 chunkID)
...
-		m_z_stream->next_out = static_cast<Bytef*>(dst);
-		m_z_stream->avail_out = m_frameSize;
-
-		int status = inflate(m_z_stream, Z_FINISH);
-		bool success = status == Z_STREAM_END && m_z_stream->total_out == m_frameSize;
+		if (m_uselz4){
+			LZ4_decompress_fast((char*)m_readBuffer, (char*)(dst), m_frameSize);
+			success = true;
+		}
+		else{
```
Comment (lightningterror) [inline pcsx2/CDVD/CsoFileReader.cpp:273]: Here, and above as well.

File: pcsx2/CDVD/CsoFileReader.cpp [hunk 3]
```diff
@@ -252,14 +264,21 @@ int CsoFileReader::ReadChunk(void *dst, s64 chunkID)
...
+			m_z_stream->next_in = m_readBuffer;
+			m_z_stream->avail_in = readRawBytes;
+			m_z_stream->next_out = static_cast<Bytef*>(dst);
+			m_z_stream->avail_out = m_frameSize;
+
+			int status = inflate(m_z_stream, Z_FINISH);
```
Comment (lightningterror) [inline pcsx2/CDVD/CsoFileReader.cpp:279]: Can be const.

**Comment by JoseAaronLopezGarcia** [comment]
> @JoseAaronLopezGarcia I fixed up your CPP file and struggled with Visual Studio, and got a working build, but when I load a ZSO file it just does nothing.
> 
> Here is the build - https://file.io/wsPRz35mv4M3

How are you generating ZSO files?

It's best to use this script:
https://github.com/ps2homebrew/Open-PS2-Loader/blob/master/pc/ziso.py
(requires python3 and pip install lz4)

**Comment by AhayriSG** [comment]
> > @JoseAaronLopezGarcia I fixed up your CPP file and struggled with Visual Studio, and got a working build, but when I load a ZSO file it just does nothing.
> > Here is the build - https://file.io/wsPRz35mv4M3
> 
> How are you generating ZSO files?
> 
> It's best to use this script: https://github.com/ps2homebrew/Open-PS2-Loader/blob/master/pc/ziso.py (requires python3 and pip install lz4)

https://www.psx-place.com/resources/psxiso-compression-tool.1258/
"ziso" named CLI tool.

**Comment by orcutt989** [comment]
@JoseAaronLopezGarcia @AhayriSG I used the python script, but I'll double check if the files work on a PS2. 

Last time I checked- they did, and I used the python script before with the new build of OPL with ZSO support.

Did you try the build? Did it work for you all?

File: PCSX2_suite.sln [hunk 4]
```diff
@@ -70,352 +70,720 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "d3d12memalloc", "3rdparty\d
...
 		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
 		Devel AVX2|x64 = Devel AVX2|x64
+		Devel AVX2|x86 = Devel AVX2|x86
 		Devel|x64 = Devel|x64
+		Devel|x86 = Devel|x86
 		Release AVX2|x64 = Release AVX2|x64
+		Release AVX2|x86 = Release AVX2|x86
 		Release|x64 = Release|x64
+		Release|x86 = Release|x86
```
Comment (tellowkrinkle) [inline PCSX2_suite.sln:88]: This whole file looks like a messed up merge
These look like they're from when we had an x86 build
And below looks like you duplicated everything

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp [hunk 5]
```diff
@@ -27,7 +27,7 @@
 
 #define XXH_STATIC_LINKING_ONLY 1
 #define XXH_INLINE_ALL 1
-#include "xxhash.h"
+#include "3rdparty/include/xxhash.h"
```
Comment (tellowkrinkle) [inline pcsx2/GS/Renderers/HW/GSTextureCache.cpp:30]: Why was this changed?

File: PCSX2_suite.sln (hunk 4 above)
Comment (orcutt989) [reply PCSX2_suite.sln:88]: Thanks for the feedback! It wasn't a merge. I followed the instructions here to build  https://wiki.pcsx2.net/PCSX2_Documentation/Compiling_on_Windows only to find out all the NIS files were removed and just use cmake. Do you have a suggestion?

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp (hunk 5 above)
Comment (orcutt989) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp:30]: Because lz4 has its own `xxhash.h` header file and it was conflicting. Would you have a suggestion for a different way to prevent the header file conflict?

File: PCSX2_suite.sln (hunk 4 above)
Comment (F0bes) [reply PCSX2_suite.sln:88]: You don't _need_ cmake on windows. Just double click the PCSX2_suite.sln or PCSX2_qt.sln

Comment (orcutt989) [reply PCSX2_suite.sln:88]: Right then what? Are there new build instructions?

Comment (F0bes) [reply PCSX2_suite.sln:88]: image

Comment (tellowkrinkle) [reply PCSX2_suite.sln:88]: Regardless of how it got here, you do need to fix it

Comment (orcutt989) [reply PCSX2_suite.sln:88]: Yep I'm on it. Thanks.

**Comment by orcutt989** [comment]
> @JoseAaronLopezGarcia @AhayriSG I used the python script, but I'll double check if the files work on a PS2.
> 
> Last time I checked- they did, and I used the python script before with the new build of OPL with ZSO support.
> 
> Did you try the build? Did it work for you all?

NVM it works! Cleaning up PR and adding menu support atm.

**Comment by refractionpcsx2** [comment]
Can you tidy up the project files please. You also need to update the pcsx2core project and the solution for Qt.

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp (hunk 5 above)
Comment (refractionpcsx2) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp:30]: How many files does lz4 actually need to compile? if it's not that many, you could extract the necessary files and just upload a custom setup as part of the PR instead of submoduling it, then you can link to our already included xxhash.

Comment (JoseAaronLopezGarcia) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp:30]: You can just include the lz4.c/h we use on PSP/PS2, it's heavily stripped down and has no other dependency, just the c and h files.
https://github.com/ps2homebrew/Open-PS2-Loader/tree/master/modules/isofs

**Comment by JoseAaronLopezGarcia** [comment]
One important note.
The CSO reader in PCSX2 can read CSO (and ZSO) files with arbitrary block sizes.
Having higher block sizes results in better compression.
On PS2 however, we only support 2048 blocks due to memory constraint (ZSO decompression runs on the IOP :P ).
So for standarization purposes, and to make sure your ZSO files will work on both real hardware and emulator, make sure you use blocks of 2048 bytes in size.

The python script already uses this block size, and doesn't really let you change it, but maxcso will by default use arbitrary block sizes, and this causes the PS2 (OPL) to not read them. You can specify the block size in the parameters to maxcso.

**Review by lightningterror — CHANGES_REQUESTED** [review]
Formatting still needs to be fixed.

**Comment by orcutt989** [comment]
> Formatting still needs to be fixed.

Thanks I havent addressed it yet. Once its fixed I will resolve your comments.

I read through the formatting guide and I'm not sure what specifically is wrong with the formatting. Could you please be more specific?

Do you all have some sort of plugin that can automate the formatting?

**Comment by tellowkrinkle** [comment]
> I read through the formatting guide and I'm not sure what specifically is wrong with the formatting. Could you please be more specific?

newlines before `{`s

> Do you all have some sort of plugin that can automate the formatting?

If you have `clang-format` installed, you can run `git clang-format HEAD~4` to format everything
Though I'd much prefer it if you squashed formatting changes onto each commit, in which case you should `git rebase -i master`, choose edit for the four commits, and `git clang-format HEAD~` after each one.

**Comment by refractionpcsx2** [comment]
What's the status on this? Also need a rebase.

**Comment by ElGranSmokio** [comment]
@refractionpcsx2 Is it really not possible to merge this PR? ZSO is the only compressed format supported by PS2's OPL and having it on PCSX2 as well would be really great since you wouldn't need to keep different files for the same game.

**Comment by AKuHAK** [comment]
> Is it really not possible to merge this PR? ZSO is the only compressed format supported by PS2's OPL and having it on PCSX2 as well would be really great since you wouldn't need to keep different files for the same game.

OP decided to close this pr, not pcsx2 team. It needs massive rebase and lz4 lib clean.
